onnxruntime/core/providers/coreml/coreml_execution_provider.cc

```diff
@@ -57,7 +58,31 @@ CoreMLExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph_vie
       [&]() {
         HashValue model_hash;
         int metadef_id = metadef_id_generator_.GenerateId(graph_viewer, model_hash);
-        return MakeString(COREML, "_", model_hash, "_", metadef_id);
+        std::string user_provided_key;
+        const Graph* main_graph = &graph_viewer.GetGraph();
+        while (main_graph->IsSubgraph()) {
+          main_graph = main_graph->ParentGraph();
+        }
+        if (main_graph->GetModel().MetaData().count(kCOREML_CACHE_KEY) > 0) {
+          user_provided_key = graph_viewer.GetGraph().GetModel().MetaData().at(kCOREML_CACHE_KEY);
+          if (user_provided_key.size() > 64 ||
+              std::any_of(user_provided_key.begin(), user_provided_key.end(),
+                          [](unsigned char c) { return !std::isalnum(c); })) {
+            user_provided_key = std::to_string(std::hash<std::string>{}(user_provided_key));
```

Comment:
I would rather document that the limit is 64 and that all chars must be std::isalnum and return an error if that is not the case. 

As a user I would expect the value to be used as is. If I was debugging the cache behavior I would not expect my explicitly provided value to have changed. I would assume caching was not working at all if I found no cache directory for my hash value. I would log a GitHub Issue complaining. We would have to determine the value was too long or had invalid chars, which is non-trivial (determine caching was enabled and determine value was too long/invalid - which requires knowing the 64-char limit and the isalnum requirement). 

If you return an error the user knows why it was wrong and can directly fix it. Odds of a GitHub Issue should be way lower.

POLS - Principle of Least Surprise. 

We can also check this as a one-off outside of gen_metadef_name instead of doing it for every partition in the graph_viewer.

Note that I think this is different to gracefully disabling caching if there's a filesystem error. We and the user don't control that. They do control the value they populate the model metadata_props with.

Reply:
Sure.  Given that 'user_provided_key' error is a user error, it's fine to throw.